### PR TITLE
Implement `formatter` as parameter for Quantity.to_string method

### PR DESCRIPTION
**Pull request by Deen-dot**
(Addressing #15933 feature request)

### Description
Introduces a new parameter, `formatter`, to the `to_string` method of the Quantity class, allowing users to specify custom formatting quickly and easily. Allows for greater flexibility in formatting quantities, especially in different contexts where consistent (e.g., always using scientific notation) or custom notation is crucial as `precision` does not currently address this alone.

Example use:
```Python
In [1]: import astropy.units as u
         import numpy as np
         quantity = 123.45 * u.m
         custom_format_spec = '.4e'
         quantity.to_string(formatter=custom_format_spec)

Out[1]: '1.2345e+02 m'

In [2]: quantity.to_string(format='latex', formatter=custom_format_spec)

Out[2]: '$1.2345 \\times 10^{2} \\; \\mathrm{m}$'

In [3]: quantity.to_string(format='latex', precision=4)

Out[3]: '$123.5 \\; \\mathrm{m}$'

In [4]: quantity.to_string(precision=4)

Out[4]: '123.4500 m'

In [5]: custom_format_spec = '+.2f'
         quantity = 123.45 * u.km/u.s
         quantity.to_string(formatter=custom_format_spec) # Custom notation in differential measurements

Out[5]: '+123.45 km / s'

In [6]: custom_format_spec = '010.2f'
         quantity = 123.45 * u.s
         quantity.to_string(formatter=custom_format_spec) # Custom notation for time with padding

Out[6]: '0000123.45 s'
```
EDIT: Implementation changed from `format_spec` to `formatter`

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: docs/changes/units/16087.feature.rst
Comment (pllim) [inline docs/changes/units/16087.feature.rst:1]: Thanks!

Some of this change log, particularly the example, is better off as a "what's new" entry. But let's wait to see what other maintainers say first.

**Comment by pllim** [comment]
Would need test(s) too for this new feature.

File: docs/changes/units/16087.feature.rst
Comment (Deen-dot) [reply docs/changes/units/16087.feature.rst:1]: Sounds good! Will continue to test and refine where necessary

File: astropy/units/quantity.py [hunk 1]
```diff
@@ -1451,25 +1465,38 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
...
+        default_formatter = (
+            {"all": lambda x: f"{x:{format_spec}}"} if format_spec else None
+        )
+
+        if format is None:
+            # format_spec overwrites precision
+            if default_formatter is None:
+                if precision is None:
+                    # Use default formatting settings
+                    return f"{self.value}{self._unitstr:s}"
```
Comment (nstarman) [inline astropy/units/quantity.py:1478]:
```suggestion
            if default_formatter is None and precision is None:
                # Use default formatting settings
                return f"{self.value}{self._unitstr:s}"
```

**Review by nstarman — COMMENTED** [review]
Thanks @Deen-dot for the contribution! I think this is a good idea. It would be nice to see some examples and/or tests showing how this interacts with the python formatting mini-language `format_spec` for which this is named. E.g. `https://docs.python.org/3/library/functions.html#format` -> `Quantity.__format__` -> ...

File: astropy/units/quantity.py [hunk 2]
```diff
@@ -1451,25 +1465,42 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
...
+        if format_spec is not None:
+            # Check for format_spec prohibited operators
+            if "%" == format_spec[-1]:
+                raise ValueError("% format_spec is not supported for Quantity.")
+            # Set default formatter
+            default_formatter = {"all": lambda x: f"{x:{format_spec}}"}
+
+        if format is None:
+            # format_spec overwrites precision
+            if default_formatter is None and precision is None:
```
Comment (mhvk) [inline astropy/units/quantity.py:1479]: Why not just `if format_spec is None and precision is None`?

File: astropy/units/quantity.py [hunk 3]
```diff
@@ -1451,25 +1465,42 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
...
                 # Use default formatting settings
                 return f"{self.value}{self._unitstr:s}"
-            else:
-                # np.array2string properly formats arrays as well as scalars
-                return (
-                    np.array2string(self.value, precision=precision, floatmode="fixed")
-                    + self._unitstr
+
+            return (
```
Comment (mhvk) [inline astropy/units/quantity.py:1483]: If we change the `if` as suggested above, then here you could do the definition of `default_formatter` (which is not used elsewhere).

File: astropy/units/tests/test_quantity.py [hunk 4]
```diff
@@ -988,6 +988,80 @@ def test_to_string(self):
             f"Quantity as KMS: {qscalar.to_string(precision=3, unit=u.km / u.s)}" == res
         )
 
+        # Scientific notation with 2 decimal places
```
Comment (mhvk) [inline astropy/units/tests/test_quantity.py:991]: I think this may be a bit more readable as a separate test method that is parametrized. I'd also just test the output of `to_string()`, not its use inside an f-string.

Indeed, I'd use the same parametrization to also check that the output for different formats is OK (latex, unicode), i.e., something like,
```
@pytest.mark.parametrize("q, generic, latex, unicode", [
    (1.5e14 * u.m / u.s, "1.50e+14 m / s", "...", "..."),
    (0.1234 * u.m / u.s, ...),
    ...
])
def test_format_spec(self, q, generic, latex, unicode):
    assert q.to_string(format_spec=format_spec) == generic
    assert q.to_string('latex', format_spec=format_spec) == latex
    ...
```

File: docs/changes/units/16087.feature.rst [hunk 5]
```diff
@@ -0,0 +1,10 @@
+Add a ``format_spec`` argument to the ``to_string`` method of the ``Quantity`` class. Enables custom number formatting, especially useful for consistent notation.
```
Comment (mhvk) [inline docs/changes/units/16087.feature.rst:1]: Could you break up the line so the individual lines are no more than 80 characters?

File: docs/changes/units/16087.feature.rst [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+The argument accepts standard Python format specifiers, allowing detailed control over the output format.
+
+Example usage:
```
Comment (mhvk) [inline docs/changes/units/16087.feature.rst:4]: I think this part can be omitted - that's for the what's-new. You may also want to check the regular documentation - I don't quite recall if we discuss `.to_string()`, but if we do, it might be nice to add an example there as well.

**Review by mhvk — COMMENTED** [review]
Thanks! This generally looks good, and my comments are mostly small. What I don't quite understand yet, though (I'm in a train, so testing things myself is a bit tricky), is how this interacts with exponential notation for latex and unicode. Does the 10^{..} still happen? I think it is actually all OK, but maybe add some tests for those too.

**Comment by nstarman** [comment]
I recall @eerovaher was working on Quantity string formatting, if he wants to take a look.

**Comment by eerovaher** [comment]
I'd prefer if the new argument was called `formatter` and worked like the `formatter` argument of [`numpy.array2string()`](https://numpy.org/doc/stable/reference/generated/numpy.array2string.html). The code is already constructing a `dict` of callables and passing it to `numpy.array2string()`, so the current `format_spec` argument is just a needless layer of obfuscation.

File: astropy/units/quantity.py [hunk 7]
```diff
@@ -1451,25 +1462,39 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
             },
         }
         formats["latex_inline"] = formats["latex"]
+        default_formatter = None
```
Comment (mhvk) [inline astropy/units/quantity.py:1465]: I think this can still be moved inside the `if format_spec is None` branch. I.e., something like
```
if format is None:
    if format_spec is None:
        if precision is None:
            return ...
        default_formatter = None
    else:
        default_formatter = {"all": lambda x: format(x, format_spec)}
```
Note the change to use `format` - it should be faster and is more explicit.

File: astropy/units/tests/test_quantity.py (hunk 4 above)
Comment (mhvk) [reply astropy/units/tests/test_quantity.py]: My intention with this comment was to be sure that we test the output for all three relevant formats for all examples. I still think that is a good idea, since the three go through different paths. Hence, I unresolved this...

**Comment by mhvk** [comment]
@eerovaher - when you suggest `formatter`, do you mean the actual dict, or a callable that can be used?

I think the present design where you can pass in a format string is good enough for most cases (and one that any person who has written f-strings or C code will be more or less familiar with), though I also rather like the idea of being able to pass in a callable so one can really do what one wants. Indeed, this is why for `Column.format` (and hence also for any `.info.format`) we support putting in either a string or a callable. It might well make sense to do that here too.

This leaves the question of the name. The advantage of `format_spec` is that that is what is used by [python](https://docs.python.org/3/library/string.html#grammar-token-format-spec-format_spec) for `format` and f-strings, which I think more people will be familiar with than with the intricacies of `array2string`. But that does not indicate the ability to pass in a callable, which `formatter` would...

**Comment by eerovaher** [comment]
> @eerovaher - when you suggest `formatter`, do you mean the actual dict, or a callable that can be used?

Accepting a `dict` would make this more like `np.array2string()`, but we do assume (even if we don't enforce) that the `dtype` of a `Quantity` is float, so accepting a callable seems more reasonable and reduces the amount of boilerplate code users would have to write.
 
> I think the present design where you can pass in a format string is good enough for most cases (and one that any person who has written f-strings or C code will be more or less familiar with), though I also rather like the idea of being able to pass in a callable so one can really do what one wants. Indeed, this is why for `Column.format` (and hence also for any `.info.format`) we support putting in either a string or a callable. It might well make sense to do that here too.

Yes, worth considering.

> This leaves the question of the name. The advantage of `format_spec` is that that is what is used by [python](https://docs.python.org/3/library/string.html#grammar-token-format-spec-format_spec) for `format` and f-strings, which I think more people will be familiar with than with the intricacies of `array2string`. But that does not indicate the ability to pass in a callable, which `formatter` would...

A scalar `Quantity` can already be formatted in f-strings, so I'd be using this function when that does not work and trying to increase similarities with f-string formatting is not very helpful. However, the similarities with `np.array2string()` are worth emphasizing. For example, both the name of the `precision` argument and the fact that it gets ignored if `formatter/format_spec` is specified clearly come from `np.array2string()`.

**Comment by mhvk** [comment]
OK, thanks! @Deen-dot - do you also agree with @eerovaher's logic? It means checking whether the argument is a `callable`, only writing the `lambda` function if it is not. I'm still not 100% on the name, but think @eerovaher's argument for `formatter` is good. In the end, I think the name doesn't matter all that much, since this argument is going to be found only via the documentation anyway.

**Comment by Deen-dot** [comment]
@eerovaher @mhvk This makes sense and sounds good to me! I'll be pushing some new changes then later today.

File: astropy/units/quantity.py [hunk 8]
```diff
@@ -1452,29 +1467,50 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
...
         # Set the precision if set, otherwise use numpy default
         pops = np.get_printoptions()
-        format_spec = f".{precision if precision is not None else pops['precision']}g"
+        if format_spec is None:
+            format_spec = (
+                f".{precision if precision is not None else pops['precision']}g"
+            )
 
         def float_formatter(value):
+            if callable(formatter):
```
Comment (mhvk) [inline astropy/units/quantity.py:1507]: This is weird. Instead, I'd write:
```
if formatter is None or isinstance(formatter, str):
    if self.dtype.kind == 'c':
        def formatter(value):
             stuff for complex
    else:
        def formatter(value):
            stuff for real
```
Indeed, it looks like this would allow substantial simplification....

**Review by mhvk — COMMENTED** [review]
Great! Though still some comments. Also it seems like now you're all set up to combine the two branches, both leading to the same `array2string` call... (but have to run now, can explain later if not obvious...)

File: astropy/units/quantity.py [hunk 9]
```diff
@@ -1452,35 +1467,64 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
         }
         formats["latex_inline"] = formats["latex"]
 
-        if format not in formats:
```
Comment (mhvk) [inline astropy/units/quantity.py:1455]: Maybe just leave this test in place (i.e., just remove the `elif` clause. It is not like it will matter for speed...

File: astropy/units/quantity.py [hunk 10]
```diff
@@ -1452,35 +1467,64 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
...
-            raise ValueError(f"Unknown format '{format}'")
-        elif format is None:
-            if precision is None:
+        format_spec = formatter if isinstance(formatter, str) else None
+
+        if format is None:
+            # format_spec overwrites precision
+            if formatter is None and precision is None:
```
Comment (mhvk) [inline astropy/units/quantity.py:1474]: I'm not really worried about speed, but for clarity it might help to put this case up before `formats` is defined:
```
        if format is None and formatter is None and precision is None:
            # Fast path for default formatting settings.
            return f"{self.value}{self._unitstr:s}"
```
But up to you.

File: astropy/units/quantity.py [hunk 11]
```diff
@@ -1452,35 +1467,64 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
...
-                ),
-            )
+            else:  # Float default latex formatter
+
+                def formatter(value):
+                    return Latex.format_exponential_notation(
+                        value, format_spec=format_spec
+                    )
 
         # The view is needed for the scalar case - self.value might be float.
```
Comment (mhvk) [inline astropy/units/quantity.py:1529]: Here, you need to do also for the latex branch,
```
        if callable(formatter):
            formatter = {"all": formatter}
```
and then use `formatter=formatter` in the `array2string` call.

File: astropy/units/tests/test_quantity.py [hunk 12]
```diff
@@ -1003,6 +1003,118 @@ def test_to_string(self):
...
+                "latex",
+                r"$(2.50-1.2j) \; \mathrm{}$",
+            ),
+        ],
+    )
+    def test_formatter(self, input_quantity, formatter, format, expected_result):
+        result = input_quantity.to_string(formatter=formatter, format=format)
+        assert result == expected_result
+
+    @pytest.mark.parametrize("format_spec", ["b", "o", "x", "c", "s"])
```
Comment (mhvk) [inline astropy/units/tests/test_quantity.py:1112]: We'll need a test where `formatter` is a `dict`. I think it is fine for this to be non-exhaustive! Just one example for generic and one for latex.

File: astropy/units/quantity.py [hunk 13]
```diff
@@ -1423,6 +1427,13 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
             - 'latex_inline': Return a LaTeX-formatted string that uses
               negative exponents instead of fractions
 
+        formatter : str, callable, optional
+            The formatter to use for the value. If a string, it should be a
+            valid format specifier using Python's mini-language. If a callable,
+            it will be treated as the default formatter for all values and will
+            overwrite default Latex formatting for exponential notation and complex
+            numbers. If not provided, the default formatter will be used.
```
Comment (mhvk) [inline astropy/units/quantity.py:1435]: add "numbers. If a `dict`, it will be passed directly to `~numpy.array2string`. If not..."

**Review by mhvk — COMMENTED** [review]
Basically done, just one last real change, and a few suggestions.

p.s. I tried whether one could combine paths to use the same `array2string` call, but ended up feeling it didn't really help clarity, so fine to leave this as you have it.

**Comment by mhvk** [comment]
Also, I worry that a plain `format_spec` is not going to work for complex numbers. Could you add a test? It may be that also for `format is None`, we need to treat float and complex separately.

File: astropy/units/quantity.py [hunk 14]
```diff
@@ -1454,33 +1475,60 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
...
+        if formatter is None or isinstance(formatter, str):
+            if self.dtype.kind == "c":  # Complex default latex formatter
+
+                def formatter(value):
+                    return "({}{}i)".format(
+                        Latex.format_exponential_notation(
+                            value.real, format_spec=format_spec
+                        ),
+                        Latex.format_exponential_notation(
+                            value.imag, format_spec="+" + format_spec
```
Comment (mhvk) [inline astropy/units/quantity.py:1519]: I guess we should remove a possible "+" in `format_spec` - `"+" + format_spec.removeprefix("+")`

**Review by mhvk — COMMENTED** [review]
Super! Very happy to see that complex worked right already! I've only got one small remaining comment, for the case of someone passes in a `format_spec` that already starts with a "+". Fix suggested inline - will also need a test... (it seems python does do this automatically as well: `format(1+1j, "+.2f")` works)

**Review by nstarman — COMMENTED** [review]
I like the look of this now. Thanks @eerovaher for the insights on this feature!

File: astropy/units/quantity.py [hunk 15]
```diff
@@ -7,6 +7,8 @@
 
 from __future__ import annotations
 
+import builtins
+
 # STDLIB
```
Comment (eerovaher) [inline astropy/units/quantity.py:12]: This is yet another example of the `# STDLIB` comment actively preventing Ruff from keeping the imports organized. We really should get rid of such comments in all of `astropy`.

File: astropy/units/tests/test_quantity.py [hunk 16]
```diff
@@ -1003,6 +1003,169 @@ def test_to_string(self):
         res = "[0 1 2] (Unit not initialised)"
         assert np.arange(3).view(u.Quantity).to_string() == res
 
+    @pytest.mark.parametrize(
+        "input_value, input_unit, format_spec, expected_result, format, precision",
+        [
+            (1.5e14, "m/s", ".2e", "1.50e+14 m / s", None, None),
+            (0.123, "m/s", "0.3f", "0.123 m / s", None, None),
+            (0.000123, "m/s", ".2e", "1.23e-04 m / s", None, None),
+            (1.23456789e15, "m/s", ".2e", "1.23e+15 m / s", None, 5),
```
Comment (eerovaher) [inline astropy/units/tests/test_quantity.py:1012]: This is the only parametrization that has a `precision` different from `None`, which is suggesting that `precision` should not be a parameter for this test and it should be tested separately instead.

File: astropy/units/tests/test_quantity.py [hunk 17]
```diff
@@ -1003,6 +1003,169 @@ def test_to_string(self):
...
+        qscalar = u.Quantity(input_value, input_unit)
+        assert (
+            qscalar.to_string(formatter=format_spec, precision=precision, format=format)
+            == expected_result
+        )
+
+    @pytest.mark.parametrize(
+        "input_quantity, formatter, format, expected_result",
+        [
+            # More explicit formatting
```
Comment (eerovaher) [inline astropy/units/tests/test_quantity.py:1118]: Use [`pytest.param()` with a custom `id`](https://docs.pytest.org/en/stable/example/parametrize.html#set-marks-or-test-id-for-individual-parametrized-test) instead of writing out these comments.

**Review by eerovaher — COMMENTED** [review]
I am not doing a thorough review, I am just pointing out a few things that stood out to me. Also, do not forget to update the documentation at `docs/units/format.rst`.

File: astropy/units/quantity.py (hunk 15 above)
Comment (mhvk) [reply astropy/units/quantity.py:12]: I agree with the sentiment - so, @Deen-dot, if you can get rid of those that certainly is fine!

File: docs/units/format.rst [hunk 18]
```diff
@@ -52,8 +52,10 @@ attributes within format strings::
     '10.500 in km'
 
 This might not work well with LaTeX strings, in which case it would be better
-to use the `Quantity.to_string() <astropy.units.Quantity.to_string()>`
-method::
+to use the `Quantity.to_string() <astropy.units.Quantity.to_string()>` method.
+This method can be enhanced with the formatter parameter, providing added
+format flexibility. It allows you to specify a dict, callable, or format_spec
+string, further refining the LaTeX format representation::
```
Comment (eerovaher) [inline docs/units/format.rst:58]: We should link to the `numpy.array2string()` and `Table` column formatting documentation as much as possible and only highlight the differences here.

File: astropy/units/quantity.py [hunk 19]
```diff
@@ -1454,33 +1471,63 @@ def to_string(self, unit=None, precision=None, format=None, subfmt=None):
...
+            format_spec = (
+                f".{precision if precision is not None else pops['precision']}g"
+            )
 
-        def float_formatter(value):
-            return Latex.format_exponential_notation(value, format_spec=format_spec)
+        # Use default formatters
+        if formatter is None or isinstance(formatter, str):
+            if self.dtype.kind == "c":  # Complex default latex formatter
+                # Disallowed sign operations for complex numbers
```
Comment (mhvk) [inline astropy/units/quantity.py:1508]: Hmm, this would also change the format for the real part, which is not the idea - you just need to change it for complex only. It is actually somewhat tricky to get completely right. For instance, python does
```
In [2]: format(1+1j, "<+20.5f")
Out[2]: '+1.00000+1.00000j   '
```
So, the width argument should apply to the whole result, not each one individually.

To be honest, I'm happy to ignore that for now for `latex`, where width has less meaning anyway (all those spaces get ignored!). But we should apply the changed format to imaginary only.

**Review by mhvk — COMMENTED** [review]
Still one comment, sorry! Note that the lines I commented on do not seem to be tested - I think it would be good to ensure they are.

I must admit that the trouble with `format_spec` and latex makes me wonder if it isn't easier to adjust `Latex.format_exponential_notation` to also deal with complex...  But for another time!

**Review by mhvk — APPROVED** [review]
Sorry, I had not seen you had addressed my last comment - and made the tests quite a bit more detail too!

It all looks good now, so let's get this in. Thanks for hanging in there - it has take quite a bit of iteration, but I think the addition has improved greatly in the process! Thanks!

File: docs/units/format.rst [hunk 20]
```diff
@@ -52,8 +52,16 @@ attributes within format strings::
...
 
 This might not work well with LaTeX strings, in which case it would be better
-to use the `Quantity.to_string() <astropy.units.Quantity.to_string()>`
-method::
+to use the `Quantity.to_string() <astropy.units.Quantity.to_string()>` method.
+Just like how `table.Column() <astropy.table.Column()>` takes a format specifier
+or callable for formatting, you can also optionally specify a format via the
+`formatter` parameter with either type or additionally with a dictionary, the
+added benefit being a more flexible or tighter LaTeX output. It relies on
+:func:`numpy.array2string()` for directly handling the :param:`formatter`, which will
```
Comment (mhvk) [inline docs/units/format.rst:60]: Hmm, readthedocs complains about `:param:` here and on l.63:
```
/home/docs/checkouts/readthedocs.org/user_builds/astropy/checkouts/16087/docs/units/format.rst:54: ERROR: Unknown interpreted text role "param".
/home/docs/checkouts/readthedocs.org/user_builds/astropy/checkouts/16087/docs/units/format.rst:54: ERROR: Unknown interpreted text role "param".
```
Simplest I think is just remove that and use double backquotes (also on l.58)

File: docs/whatsnew/6.1.rst [hunk 21]
```diff
@@ -34,6 +34,32 @@ on Windows and potentially 32-bit architectures. Previously on those platforms,
...
+
+    from astropy import units as u
+    q = u.Quantity(123.456, u.m)
+    custom_format = lambda x: f"\\approx {float(x):.1f}",
+    print(q.to_string(formatter=custom_format, format='latex'))
+    # Outputs: "$\approx 123.5 \; \mathrm{m}$",
+    print(q.to_string(formatter='.3e', format='latex'))
+    # Outputs: '$1.235 \times 10^{2} \; \mathrm{m}$'
+    print(q.to_string(precision=3, format='latex'))
+    # Outputs: '$123 \; \mathrm{m}$'
```
Comment (eerovaher) [inline docs/whatsnew/6.1.rst:62]: This code example needs to be rewritten in its entirety, both to match other code examples in style and to ensure the example is actually tested.

Comment (mhvk) [reply docs/whatsnew/6.1.rst:62]: Hmm, you're right - when I saw this I had this vague memory that we do not run doctests on what's-new, but clearly we do write them as such - @Deen-dot, see https://docs.astropy.org/en/v6.0.0/whatsnew/6.0.html for examples (no `.rst` any more since we don't guarantee those stay the same, but it used to be an `.rst` file too).

Comment (eerovaher) [reply docs/whatsnew/6.1.rst:62]: It appears the What's New entries are indeed not tested. Including the entries for the next release in the tests would be a good idea, but doing that goes beyond the scope here. However, writing the code example in the same style as all the others is a good idea regardless.

Comment (pllim) [reply docs/whatsnew/6.1.rst:62]: doctestplus only picks up things with `>>>`.

**Comment by mhvk** [comment]
closing/opening to retrigger tests

**Comment by mhvk** [comment]
Hmm, circleci seems stuck - the close/open did not help. I'll press auto-merge to indicate this should go in. Thanks, @Deen-dot!

**Comment by pllim** [comment]
This needs a rebase because CI rules have changed since we ditched Python 3.9. Please rebase. Thanks!

**Comment by Deen-dot** [comment]
Woops! I messed up a little with the GitKraken push, working to resolve (please ignore for now)

**Comment by mhvk** [comment]
@Deen-dot - sorry but could you rebase one more time? I'm pretty sure the readthedocs failures were unrelated (and hence solved), but we might as well check.

**Comment by Deen-dot** [comment]
@mhvk Seems like all check issues were resolved 👍
